pkg/providers/tinkerbell/pbnj/client.go

```diff
@@ -37,16 +47,67 @@ func NewPBNJClient(pbnjGrpcAuth string) (*Pbnj, error) {
 	return pbnjObj, nil
 }
 
-func (p *Pbnj) ValidateBMCSecretCreds(ctx context.Context, bmcInfo BmcSecretConfig) error {
-	_, err := p.pbnj.MachinePower(ctx, p.NewPowerRequest(bmcInfo))
+func (p *Pbnj) GetPowerState(ctx context.Context, bmcInfo BmcSecretConfig) (PowerState, error) {
+	status, err := p.ExecutePowerRequest(ctx, NewPowerRequest(bmcInfo, v1.PowerAction_POWER_ACTION_STATUS))
+	if err != nil {
+		return PowerStateUnknown, err
+	}
+
+	result := strings.ToLower(status.Result)
+
+	if strings.Contains(result, string(PowerStateOn)) {
+		return PowerStateOn, nil
+	} else if strings.Contains(result, string(PowerStateOff)) {
+		return PowerStateOff, nil
+	}
+
+	return PowerStateUnknown, nil
+}
+
+func (p *Pbnj) PowerOff(ctx context.Context, bmcInfo BmcSecretConfig) error {
+	powerState, err := p.GetPowerState(ctx, bmcInfo)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get power state when trying to power off machine: %s %v", bmcInfo.Host, err)
+	}
+
+	// if power state is already off pbnj throws an error so we only issue the call if not already in off state
```

Comment:
Document the race condition. If a machine, for whatever reason, powers off _after_ the `GetPowerState()` but before `ExecutePowerRequest()` then we'll receive an error. It would be useful to maintainers to explicitly point that out in a short comment.

Reply:
Alternatively, given this is a gRPC call, you could do what we talked about and check the contents of the error. With gRPC errors are free-form text so its brittle but unless the upstream instruments an error payload with more data there's not much we can do.

Reply:
Yea I would do that. Combined with a clear error message it'll be clear pretty quickly that the code is looking for the wrong error message because it'll be erroring all the time. 

We probably should log when we swallow the error though. You'll want to inject a logger and log at a reasonably high verbosity level - I don't know what the teams current approach is to defining levels.

Reply:
Ok I pushed the update. Also after furthre testing I found error only occurs if powered off. On powered on if you issue a power on command ipmi does not throw an error.